src/Chainweb/Pact/Backend/RelationalCheckpointer.hs

```diff
@@ -279,31 +282,52 @@ doGetBlockHistory dbenv blockHeader d = runBlockEnv dbenv $ do
     startTxId <- if (bHeight == genesisHeight v cid)
       then pure 0  -- genesis block
       else getEndTxId db (pred bHeight) (_blockParent blockHeader)
-    history <- queryHistory db (domainTableName d) startTxId endTxId
-    return $! BlockTxHistory $ foldl' groupByTxid mempty history
+    let tname = domainTableName d
+    history <- queryHistory db tname startTxId endTxId
+    let (!hkeys,tmap) = foldl' procTxHist (S.empty,mempty) history
+    !prev <- M.fromList . catMaybes <$> mapM (queryPrev db tname startTxId) (S.toList hkeys)
+    return $ BlockTxHistory tmap prev
   where
     v = _blockChainwebVersion blockHeader
     cid = _blockChainId blockHeader
     bHeight = _blockHeight blockHeader
 
-    groupByTxid :: Ord a => M.Map a [b] -> (a,b) -> M.Map a [b]
-    groupByTxid r (t,l) = M.insertWith (++) t [l] r
+    procTxHist
+      :: (S.Set Utf8, M.Map TxId [TxLog Value])
+      -> (Utf8,TxId,TxLog Value)
+      -> (S.Set Utf8,M.Map TxId [TxLog Value])
+    procTxHist (ks,r) (uk,t,l) = (S.insert uk ks, M.insertWith (++) t [l] r)
 
     -- Start index is inclusive, while ending index is not.
     -- `endingtxid` in a block is the beginning txid of the following block.
-    queryHistory :: Database -> Utf8 -> Int64 -> Int64 -> IO [(TxId,TxLog Value)]
+    queryHistory :: Database -> Utf8 -> Int64 -> Int64 -> IO [(Utf8,TxId,TxLog Value)]
     queryHistory db tableName s e = do
       let sql = "SELECT txid, rowkey, rowdata FROM [" <> tableName <>
                 "] WHERE txid >= ? AND txid < ?"
       r <- qry db sql
            [SInt s,SInt e]
            [RInt,RText,RBlob]
       forM r $ \case
-        [SInt txid, SText key, SBlob value] -> (fromIntegral txid,) <$> toTxLog d key value
+        [SInt txid, SText key, SBlob value] -> (key,fromIntegral txid,) <$> toTxLog d key value
         err -> internalError $
-               "readHistoryResult': Expected single row with three columns as the \
+               "queryHistory: Expected single row with three columns as the \
                \result, got: " <> T.pack (show err)
 
+    -- Get last tx data, if any, for key before start index.
+    queryPrev :: Database -> Utf8 -> Int64 -> Utf8 -> IO (Maybe (RowKey,TxLog Value))
+    queryPrev db tableName s k@(Utf8 sk) = do
+      let sql = "SELECT rowdata FROM [" <> tableName <>
+                "] WHERE rowkey = ? AND txid < ? " <>
+                "ORDER BY txid DESC LIMIT 1"
+      r <- qry db sql
+           [SText k,SInt s]
+           [RBlob]
+      case r of
+        [] -> return Nothing
+        [[SBlob value]] -> Just . (RowKey $ toS sk,) <$> toTxLog d k value
+        _ -> internalError $ "queryPrev: expected 0 or 1 rows, got: " <> T.pack (show r)
+
+
```

Comment:
instead of querying the database twice, would it be possible to just condense this query into a single one, collecting just the first two rows generated by a rowkey for some table ordered by txid?

Reply:
@slpopejoy thoughts?

Reply:
I banged my head against that very question and came up empty, I think at a minimum you need a temp table. Note that strategically this stuff will all be in chainweb-data at some point so less-optimum is ok for now.